salt/modules/zabbix.py

```diff
@@ -0,0 +1,1152 @@
+
+
+def _login(**kwargs):
+    '''
+    Log in to the API and generate the authentication token.
+
+    Returns:
+        On success connargs dictionary with auth token and frontend url,
+        False on failure.
+
+    '''
+    connargs = dict()
+
+    def _connarg(name, key=None):
+        '''
+        Add key to connargs, only if name exists in our kwargs or, as zabbix.<name> in __opts__ or __pillar__
+
+        Evaluate in said order - kwargs, opts, then pillar. To avoid collision with other functions,
+        kwargs-based connection arguments are prefixed with 'connection_' (i.e. '_connection_user', etc.).
+
+        Inspired by mysql salt module.
+        '''
+        if key is None:
+            key = name
+
+        if name in kwargs:
+            connargs[key] = kwargs[name]
+        else:
+            prefix = '_connection_'
+            if name.startswith(prefix):
+                try:
+                    name = name[len(prefix):]
+                except IndexError:
+                    return
+            val = __salt__['config.option']('zabbix.{0}'.format(name), None)
+            if val is not None:
+                connargs[key] = val
+
+    _connarg('_connection_user', 'user')
+    _connarg('_connection_password', 'password')
+    _connarg('_connection_url', 'url')
+
+    if 'url' not in connargs:
+        connargs['url'] = _frontend_url()
+
+    try:
+        if connargs['user'] and connargs['password'] and connargs['url']:
+            params = {'user': connargs['user'], 'password': connargs['password']}
+            method = 'user.login'
+            ret = _query(method, params, **connargs)
+            auth = ret['result']
+            connargs['auth'] = auth
+            connargs.pop('user', None)
+            connargs.pop('password', None)
+            return connargs
+        else:
+            raise KeyError
+    except KeyError:
+        return False
+
+
+def _params_extend(params, _ignore_name=False, **kwargs):
+    '''
+    Extends the params dictionary by values from keyword arguments.
+
+    Args:
+        params: Dictionary with parameters for zabbix API.
+        _ignore_name: Salt State module is passing first line as 'name' parameter. If API uses optional parameter
+                      'name' (for ex. host_create, user_create method), please use 'visible_name' or 'firstname'
+                      instead of 'name' to not mess these values.
+
+    Returns:
+        Extended params dictionary with parameters.
+
+    '''
+    for key in kwargs.keys():
+        if not key.startswith('_'):
+            params.setdefault(key, kwargs[key])
+
+    # ignore name parameter passed from Salt state module
+    if _ignore_name:
+        params.pop('name', None)
+        if 'firstname' in params:
+            params['name'] = params.pop('firstname')
+        elif 'visible_name' in params:
+            params['name'] = params.pop('visible_name')
+
+    return params
+
+
+def apiinfo_version(**connection_args):
```

Comment:
This usage of kwargs makes it somewhat difficult to understand how to utilize this function. If possible, it might be better to specify the parameters explicitly. If not, I think we need to document usage in the docstring more thoroughly here.